opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
```

Comment:
> The only commonality among them from OT point of view is that they deal with a text protocol, so OT provides a way to encode span into virtual "headers" as string key/value pairs.

This, and the byte stream, and the decision as to which to use just doesn't gel with me.

You've taken a Span, then transformed it to a k/v, then handled to the RPC framework with the presumption the map is the right thing for it to work with. That's presumptuous and creating an API that doesn't evolve well. 

On the other hand if you take the SpanExtractor and work with the problem of headers coming out (same thing really)

    public interface SpanExtractor {
        Span getSpan();
    }
Then add on to that the use-case for the commonality you expect you get (something like)

    public class SpanFromMapExtractor {
        public SpanFromMapExtractor(Map<String,String> spanIds, Map<String,String> headers) { ..}
        public Span getSpan() {
            return new Span(spanIds.get("trace_id"), spanIds.get("span_id"), headers);
    }
You get the picture. Here you haven't forced the presumption of the text or byte array or map up to the top level of the api but kept it within suggested helper classes. From here the API can evolve to offer SpanFromServletExtractor, SpanFromThriftExtractor, SpanFromSpringExtractor, etc. And from these classes you get to start offering more the nuts&bolts stuff like optimisations to skip reading headers while extracting, without polluting the top-level design. These helper classes are easy to depose of as well over time, or move into their corresponding frameworks even. End-user developers also have full access to this functionality, again without complicating the initial API we expose them to.
> The actual http libraries all have different APIs for http request class…

And that too can be better designed for, as pointed out in the SpanExtractor example.

> opentracing-python-instrumentation/opentracing_instrumentation/http_client.py#L59

Yes that is end-user code, and I can see a better design there possible that steers away from byte streams and k/v maps, and other presumptions about what RPC frameworks offer you.